File: Algorithm/Sort_Algorithm/go/mergesort01/MergeSort.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

func merge(a []int, b []int) []int {
	var result []int
	i := 0
	j := 0
	// 对比A数组与B数组，把小的值放到结果集里
	for i < len(a) && j < len(b) {
		if a[i] <= b[j] {
			result = append(result, a[i])
			i++
		} else {
			result = append(result, b[j])
			j++
		}
	}
	// 把剩余的元素放到结果集中
	if i < len(a) {
		result = append(result, a[i:]...)
	} else if j < len(b) {
		result = append(result, b[j:]...)
	}
	return result
}

func mergePass(ab []int) []int {
	if len(ab) == 1 {
		return ab
	}
	m := len(ab) / 2
	return merge(mergePass(ab[:m]), mergePass(ab[m:]))
}

func main() {
	array := make([]int, 1*1000*1000)
	for i := range array {
		array[i] = rand.Int()
	}

	startTime := time.Now()

	newArray := mergePass(array)

	fmt.Println("done:", time.Now().Sub(startTime))

	for i := 0; i < 100; i++ {
		fmt.Println(newArray[i])
	}

}

File: Algorithm/archived/Sort_Algorithm/lib/MergeSort.js
'use strict'

module.exports = (sr) => {
	let tr = new Array(sr.length)
	let k = 1
	while (k < sr.length) {
		mergePass(sr, tr, k, sr.length - 1)
		k *= 2
		mergePass(tr, sr, k, sr.length - 1)
		k *= 2
	}
}

function mergePass(sr, tr, s, n) {
	let i
	let j
	for (i = 0; i <= n - 2 * s - 1; i += 2 * s) {
		merge(sr, tr, i, i + s - 1, i + 2 * s - 1)
	}
	if (i < n - s + 1) {
		merge(sr, tr, i, i + s - 1, n)
	} else {
		for (j = i; j <= n; j++) {
			tr[j] = sr[j]
		}
	}
}

function merge(sr, tr, i, m, n) {
	let j, k, l
	for (j = m + 1, k = i; i <= m && j <= n; k++) {
		tr[k] = sr[i] < sr[j] ? sr[i++] : sr[j++]
	}
	if (i <= m) {
		for (l = 0; l <= m - i; l++) {
			tr[k + l] = sr[i + l]
		}
	}
	if (j <= n) {
		for (l = 0; l <= n - j; l++) {
			tr[k + l] = sr[j + l]
		}
	}
}
